File: Problems/No.1971_Find_if_Path_Exists_in_Graph/find_if_path_exists_in_graph.cpp
class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        
        // parameters
        std::vector<std::vector<int>> edge_list(n);
        std::vector<bool> arrivable_nodes(n, false);
        std::queue<int> node2go;
        int curr;

        // start computing
        // collate edges
        for (auto &edge: edges) {

            edge_list[edge[0]].push_back(edge[1]);
            edge_list[edge[1]].push_back(edge[0]);

        }

        // start visiting
        node2go.push(source);
        arrivable_nodes[source] = true;
        while (!node2go.empty()) {

            curr = node2go.front();
            node2go.pop();
            
            if (curr == destination) return true;
            else {

                for (auto &node: edge_list[curr]) {
                    if (!arrivable_nodes[node]) {

                        arrivable_nodes[node] = true; 
                        node2go.push(node);

                    }
                }

            }

        }

        return false;

    }
};

File: Problems/No.1971_Find_if_Path_Exists_in_Graph/find_if_path_exists_in_graph.py
class Solution(object):
    def validPath(self, n, edges, source, destination):
        """
        :type n: int
        :type edges: List[List[int]]
        :type source: int
        :type destination: int
        :rtype: bool
        """

        # parameters
        edgelist = [[] for _ in range(n)]
        arrivable_nodes = [False]*n
        node2go = []
        curr = None
        
        # start computing
        # collate edges
        for edge in edges:
            edgelist[edge[0]].append(edge[1])
            edgelist[edge[1]].append(edge[0])
        
        # start visiting
        node2go.append(source)
        arrivable_nodes[source] = True
        while node2go:
            curr = node2go.pop(0)
            if (curr == destination):
                return True
            else:
                for node in edgelist[curr]:
                    if not arrivable_nodes[node]:
                        arrivable_nodes[node] = True
                        node2go.append(node)
        
        return False
